src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -3297,20 +3299,43 @@ protected override IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol
                 throw new ArgumentException(CSharpResources.OutIsNotValidForReturn);
             }
 
+            if (callingConvention != SignatureCallingConvention.Unmanaged && !callingConventionTypes.IsDefaultOrEmpty)
+            {
+                throw new ArgumentException(string.Format(CSharpResources.CallingConventionTypesRequireUnmanaged, nameof(callingConventionTypes), nameof(callingConvention)));
+            }
+
+            if (!callingConvention.IsValid() || callingConvention == SignatureCallingConvention.VarArgs)
+            {
+                throw new ArgumentOutOfRangeException(nameof(callingConvention));
+            }
+
             var returnTypeWithAnnotations = TypeWithAnnotations.Create(returnType.EnsureCSharpSymbolOrNull(nameof(returnType)), returnType.NullableAnnotation.ToInternalAnnotation());
             var parameterTypesWithAnnotations = parameterTypes.SelectAsArray(
                 type => TypeWithAnnotations.Create(type.EnsureCSharpSymbolOrNull(nameof(parameterTypes)), type.NullableAnnotation.ToInternalAnnotation()));
+            var internalCallingConvention = callingConvention.FromSignatureConvention();
+            var conventionModifiers = internalCallingConvention == CallingConvention.Unmanaged && !callingConventionTypes.IsDefaultOrEmpty
+                ? callingConventionTypes.SelectAsArray((type, i) => getCustomModifierForType(type, i))
+                : ImmutableArray<CustomModifier>.Empty;
 
-            // PROTOTYPE(func-ptr): Allow passing in calling convention and custom modifiers for return/parameters
             return FunctionPointerTypeSymbol.CreateFromParts(
-                Cci.CallingConvention.Default,
+                callingConvention.FromSignatureConvention(),
+                conventionModifiers,
                 returnTypeWithAnnotations,
-                refCustomModifiers: default,
-                returnRefKind,
-                parameterTypesWithAnnotations,
-                parameterRefCustomModifiers: default,
-                parameterRefKinds,
-                this).GetPublicSymbol();
+                returnRefKind: returnRefKind,
+                parameterTypes: parameterTypesWithAnnotations,
+                parameterRefKinds: parameterRefKinds,
+                compilation: this).GetPublicSymbol();
+
+            static CustomModifier getCustomModifierForType(INamedTypeSymbol type, int index)
+            {
+                var internalType = type.EnsureCSharpSymbolOrNull($"{nameof(callingConventionTypes)}[{index}]");
```

Comment:
>EnsureCSharpSymbolOrNull [](start = 40, length = 24)

It looks like this check doesn't guarantee that the types come from this compilation core library, just from a core library. In other context we control where the symbols are coming from, but not for inputs to CommonCreateFunctionPointerTypeSymbol API. Please add a test which feeds types from an unrelated core library to CreateFunctionPointerTypeSymbol API.